crates/red_knot/src/symbols.rs

```diff
@@ -410,6 +457,27 @@ where
     }
 }
 
+pub(crate) struct ParentScopeIdsIterator<'a> {
+    table: &'a SymbolTable,
+    scope_id: Option<ScopeId>,
+}
+
+impl<'a> Iterator for ParentScopeIdsIterator<'a> {
+    type Item = ScopeId;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(scope_id) = self.scope_id {
+            let parent = self.table.scopes_by_id[scope_id].parent;
+            self.scope_id = parent;
+            parent
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a> FusedIterator for ParentScopeIdsIterator<'a> {}
+
 pub(crate) struct ScopeIterator<'a, I> {
```

Comment:
It used to return `Scope` objects, now it returns `(ScopeId, Scope)` pairs, since a `Scope` doesn't know its own id. I think this is a pretty natural pair and I don't really want to introduce a new struct for it, but I could document it.

Maybe we don't need this iterator at all if we just implement all access of `Scope` data via methods exposed on `ScopeId` that take a `SymbolTable`. Then `Scope` becomes an internal implementation detail. The advantage here is that you always just work with `ScopeId` and don't worry about `ScopeId` vs `Scope`. The disadvantage is that you can only access data one field at a time, and if you need multiple you end up doing multiple lookups. But considering these lookups are just indexed access into an `IndexVec`, maybe they are so cheap this doesn't matter.